sig-and-wg/sig-core/proposals/common-validator.md

```diff
@@ -0,0 +1,266 @@
+TAG = $(DOCKER_TAG)
+
+.PHONY: ci-pr ci-master ci-release resolve build test format lint build-image push-image
+
+ci-pr: validate build-image push-image
+ci-master: ci-pr
+ci-release: ci-master
+
+validate: resolve build test format lint
+
+define TARGETS
+resolve:
+	dep ensure -v -vendor-only
+build:
+	go build -o $(APP_NAME) $(1)
+test:
+	go test ./...
+format:
+	go fmt
+lint:
+	go vet
+endef
+
+build-image:
+	docker build -t $(APP_NAME) .
+push-image:
+	docker tag $(APP_NAME) $(IMG):$(TAG)
+	docker push $(IMG):$(TAG)
+```
+
+`ui-api-layer`:
+```makefile
+APP_NAME = ui-api-layer
+REPOSITORY_PATH = $(realpath $(shell pwd)/../..)
+include $(REPOSITORY_PATH)/common//template.mk
+
+$(eval $(call TARGETS,main.go))
+```
+
+Thanks to `template.mk` we will have one place in repository with targets definitions and it will be much easier to maintain it.
+
+>**NOTE**: All components needs to be build by CI system in case of changes in `template.mk`.
+
+### Sandbox Validation
+
+Developers often have different environment than CI system. For example different version of Golang. It is important to provide a way for verifying code against CI environment. For that lets create another target `validate-sandbox`, that will execute all verification in same `buildpack` Docker image as CI system is using.
+
+Definition of `validate-sandbox` target in `template.mk`:
+```makefile
+COMPONENT_REL_PATH=$(shell echo $(shell pwd) | sed 's,$(REPOSITORY_PATH)/,,g')
+
+validate-sandbox:
+	docker run --rm -v "$(REPOSITORY_PATH):/workspace/go/src/github.com/kyma-project/kyma" \
+	--workdir "/workspace/go/src/github.com/kyma-project/kyma" \
+	eu.gcr.io/kyma-project/prow/test-infra/buildpack-golang:v20181119-afd3fbd \
+	make -C "/workspace/go/src/github.com/kyma-project/kyma/$(COMPONENT_REL_PATH)" validate
+```
+
+## Summary
+
+To unify our validation and `makefile` targets we should define a template that will be included in all components `makefiles`.
+
+The definition of template `makefile` may looks like that:
+```makefile
+IMG = $(DOCKER_PUSH_REPOSITORY)$(DOCKER_PUSH_DIRECTORY)/$(APP_NAME)
+TAG = $(DOCKER_TAG)
+COMPONENT_REL_PATH=$(shell echo $(shell pwd) | sed 's,$(REPOSITORY_PATH)/,,g')
+
+.PHONY: ci-pr ci-master ci-release resolve build test format lint build-image push-image
+
+ci-pr: validate build-image push-image
+ci-master: ci-pr
+ci-release: ci-master
+
+validate: resolve build test format lint
+
+validate-sandbox:
+	docker run --rm -v "$(REPOSITORY_PATH):/workspace/go/src/github.com/kyma-project/kyma" \
+	--workdir "/workspace/go/src/github.com/kyma-project/kyma" \
+	eu.gcr.io/kyma-project/prow/test-infra/buildpack-golang:v20181119-afd3fbd \
+	make -C "/workspace/go/src/github.com/kyma-project/kyma/$(COMPONENT_REL_PATH)" validate
+
+define TARGETS
+resolve:
+	dep ensure -v -vendor-only
+build:
+	go build -o $(APP_NAME) $(1)
+test:
+	go test ./...
+format:
+	go fmt
+lint:
+	go vet
+endef
+
+build-image:
```

Comment:
In general the proposal looks good to me. 

I Just want to mention that in the event-bus component, we are building/pushing three docker images, so we have to override the `build-image` and `push-image` recipes in the event-bus makefile. Will this cause any issues or concerns?

Reply:
I'm aware of that, and I will update `template.mk` to handle such situation. I will ping you when I update it 😄

Reply:
@marcobebway please keep in mind that there is also a separate proposal from @a-thaler that is already accepted by many people that assumes split of event-bus directory. So anyway it should not be a problem. Would be actually an awesome stuff in for 0.8 scope we could go forward with implementation of this proposal and also cleanup of directories structure/name

Reply:
@marcobebway Proposal updated, but as @derberg mentioned, we should consider if we still want it 😄